### PR TITLE
QuadratureFunction::Integrate

### PR DESCRIPTION
**Pull request by pazner**
Since quadrature spaces and functions naturally have to do with quadrature, it makes sense to be able to integrate them. This PR adds:

* `QuadratureSpaceBase::GetWeights`, which returns the integration weights (with mesh Jacobians) at every point in the quadrature space
* `QuadratureSpaceBase::Integrate`, which integrates Coefficients (or VectorCoefficients) over a quadrature space
* `QuadratureFunction::Integrate`, which returns the integral of a quadrature function

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4059](https://github.com/mfem/mfem/pull/4059) | @pazner | @tzanio | @rcarson3 + @tarikdzanic | 1/16/24 | 2/20/24 | 2/21/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/qspace.cpp
Comment (rcarson3) [inline fem/qspace.cpp:63]: This will return the wrong results if the Mesh nodes have been updated or the GeometricFactors have been updated, but the internal `QuadratureSpace` values have not been updated. The easiest check to ensure validity would be to compare the pointer of the GeometricFactor class used in the last calculation to current one stored on the mesh. If they pointers are equal then the weights are the same, but if they are different then you should need to reconstruct the weights: https://github.com/mfem/mfem/blob/master/mesh/mesh.hpp#L1730-1747

File: fem/qfunction.cpp
Comment (rcarson3) [inline fem/qfunction.cpp:284]: This I believe is more of an issue with MFEM lacking a way to easily calculate reductions more than anything, but the above really shouldn't require constructing a whole new QF for this reduction. I would even argue that the even the `QuadratureSpaceBase::weights` variable is also not required here but that's only the case if there was a nice reduction operator kinda like RAJA's `RAJA::ReduceSum<ExecType, double>` type (although their current implementation has limitations...). Since, you can calculate everything required on the fly...

File: fem/qspace.cpp
Comment (rcarson3) [reply fem/qspace.cpp:63]: I'd prefer to see this check added just as I do think this is an issue users will run into especially as there is no external way for them to update the weights themselves.

File: fem/qspace.hpp
Comment (rcarson3) [inline fem/qspace.hpp:111]: Might be a note for future updates but these methods and the `QuadratureFunction::Integrate(Vector...)` one create a number of potentially large temporary memory objects. It might be useful to utilize the work in https://github.com/mfem/mfem/pull/4065 here.

File: fem/qfunction.cpp
Comment (rcarson3) [inline fem/qfunction.cpp:260]: I know this isn't relevant to this PR, but I just want to say that Vector::operator*() really makes initially reviewing this code harder. It would have been nice if it had been called `InnerProduct` instead just for code readability. It took me looking up the Vector documentation to understand what the code was actually doing.

**Review by rcarson3 — CHANGES_REQUESTED** [review]
I have a few comments more about how things could be greatly simplified if MFEM had support for reduction operators like RAJA's reduce types, but largely this PR appears to do what it says.

Outside that I did comment on some changes I'd like to see made in-order to prevent issues that users will run into.

File: fem/qspace.cpp
Comment (pazner) [reply fem/qspace.cpp]: Really good point, thanks for noticing that.

In my testing, comparing the pointer wasn't robust (sometimes the same memory address was used for the new geometric factors, so the weights weren't refreshed).

Instead I added `Mesh::nodes_sequence`, analogous to `Mesh::sequence`, to know when the cache has been invalidated (commits 6d35c65f954d11a8c7f66dd9c3baf361c11f0645 and 8d234b5358e9f2306116deee4866032aad5a233a).

File: fem/qfunction.cpp
Comment (pazner) [reply fem/qfunction.cpp]: Good point. There actually is a free function called `InnerProduct`, I switched to that instead of `operator*` in 4bfe05a9a5fd7788e022346eaf56c1a6354cc50c.

File: fem/qspace.hpp
Comment (pazner) [reply fem/qspace.hpp:112]: Yes, definitely. We should make note of this, so then when something like #4065 is merged, we take advantage of it here. (This is exactly the intended use-case).

File: fem/qfunction.cpp
Comment (pazner) [reply fem/qfunction.cpp]: Yeah, as far as I know, MFEM only really supports GPU reductions for dot products or sums. In this case, it could be modeled as a short and wide matrix times a vector, or as a kind of strided reduction… are there good ways of doing those kinds of reductions without a portability layer like RAJA? Would something like `cublasDgemv` be efficient? Or even `cublasDdot` takes a stride that could be used to eliminate the temporary. It would just be a little more boilerplate to handle HIP compatibility, etc.

File: fem/qspace.cpp
Comment (rcarson3) [reply fem/qspace.cpp]: Thanks for adding the above fix!

File: fem/qfunction.cpp
Comment (rcarson3) [reply fem/qfunction.cpp]: If we added some iterators to `QuadratureFunction` then you could use `std::accumulate` or the CUDA Thrust / rocThrust equivalent `thrust::reduce` operators to do this sorta thing.  However, I think moving to something like that would probably be a bit overkill for this sorta thing... I think with the #4065 merge then the temporary object creation is much less of an issue. I wouldn't worry about trying to fix this as the above though as there likely won't be a none abstraction layer equivalent available for all our platforms for a bit. 

As a side note, I did hear at SC23 that this will get better as apparently in C++26(?) there's supposedly going to be some better parallel execution aspects added in to the stdlib that's targeted more towards GPU vendors / some later patch changes into LLVM which will supposedly allow vendors to start allowing certain functions be host/device compatible.

**Review by rcarson3 — APPROVED** [review]
With the latest changes I think this is fine for now. We've noted as a future PR that once #4065 has been merged then a lot of the temp objects could utilize that allocator which fixes my issues with the number of temp objects being created.

File: fem/qspace.cpp
Comment (pazner) [reply fem/qspace.cpp]: @tzanio, @v-dobrev, you might want to take a look at the addition of `Msh::nodes_sequence` to see if that makes sense. Related to #3343.

**Comment by tzanio** [comment]
ping: @tarikdzanic

**Review by tarikdzanic — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

File: fem/qspace.cpp
Comment (v-dobrev) [reply fem/qspace.cpp]: I think the field `Mesh::nodes_sequence` should be added to `Mesh::Swap` and probably other methods of this type.

I noticed a call `GetIntRule(0)` in `QuadratureSpaceBase::ConstructWeights()` -- it will be good to guard that for empty local meshes.

**Comment by tzanio** [comment]
Waiting on response to https://github.com/mfem/mfem/pull/4059#discussion_r1486942237

File: fem/qspace.cpp
Comment (pazner) [reply fem/qspace.cpp]: The empty meshes are handled in 4df15bf52605c3c84329e8b4718dcbda219ab51d.

I made some changes to handle `nodes_sequence` in d461ee1bca16d6c6817e581d7c48185fc2101893. I think it's OK, but not 100% sure I covered everything properly…

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Pull request merged**
